Kubernetes v1.35 Announcement Blog Post (#53498)
kubernetes/website: content/en/blog/_posts/2025-12-17-kubernetes-v1-35-release/index.md

The goal is to make `TrafficDistribution` clearer and less ambiguous while enabling explicit same-node routing. `PreferClose` will remain in the API but is no longer the recommended value.

This work was done as part of [KEP #3015](https://kep.k8s.io/3015) led by SIG Network.

### Job API managed-by mechanism

This KEP introduces support for the `managedBy` field to let an external controller handle Job status synchronization. The feature is driven by [MultiKueue](https://github.com/kubernetes-sigs/kueue/tree/main/keps/693-multikueue), a multi-cluster Job dispatching system where a Job created in a management cluster is mirrored and executed in a worker cluster, with status propagated back. To enable this workflow, the built-in Job controller must be disabled so that the Kueue controller can manage status updates instead.
The goal is to allow clean delegation of Job synchronization to another controller. It does not aim to pass custom parameters to that controller or modify CronJob concurrency policies.

This work was done as part of [KEP #4368](https://kep.k8s.io/4368) led by SIG Apps.

> ### DRA: structured parameters

Reviewer: This whole paragraph likely not needed in release blog

> ### DRA: structured parameters

Reviewer 2: LGTM

This work was done as part of [KEP #127](https://kep.k8s.io/127) led by SIG Node.

### VolumeSource: OCI artifact and/or image

When creating a Pod, you often need to provide data, binaries, or configuration files for your containers. This meant including the content into the main container image or using a custom init container to download and unpack files into an `emptyDir`. Both these approaches are still valid. Kubernetes v1.31 added support for the `image` volume type allowing Pods to declaratively pull and unpack OCI container image artifacts into a volume. This lets you package and deliver data-only artifacts such as configs, binaries, or machine learning models using standard OCI registry tools. 
With this feature, you can fully separate your data from your container image and remove the need for extra init containers or startup scripts. The image volume type has been in beta since v1.33 and will be enabled by default in v1.35.

This work was done as part of [KEP #4639](https://kep.k8s.io/4639) led by SIG Node.

> ### Expose node topology labels via Downward API

Reviewer: maybe in release notes it needs to be in highlights

* [Pod Generation](https://kep.k8s.io/5067)
* [Invariant Testing](https://kep.k8s.io/5468)
* [In-Place Update of Pod Resources](https://kep.k8s.io/1287)
* [Fine-grained SupplementalGroups control](https://kep.k8s.io/3619)
* [DRA: structured parameters](https://kep.k8s.io/4381)
* [Structured Authentication Config](https://kep.k8s.io/3331)
* [Add support for a drop-in kubelet configuration directory](https://kep.k8s.io/3983)
* [Remove gogo protobuf dependency for Kubernetes API types](https://kep.k8s.io/5589)
* [kubelet image GC after a maximum age](https://kep.k8s.io/4210)
* [Kubelet limit of Parallel Image Pulls](https://kep.k8s.io/3673)
* [Comparable Resource Version](https://kep.k8s.io/5504)
* [Add a TopologyManager policy option for MaxAllowableNUMANodes](https://kep.k8s.io/4622)
* [Include kubectl command metadata in http request headers](https://kep.k8s.io/859)
* [PreferSameNode Traffic Distribution (formerly PreferLocal traffic policy / Node-level topology)](https://kep.k8s.io/3015)
* [Job API managed-by mechanism](https://kep.k8s.io/4368)
* [Transition from SPDY to WebSockets](https://kep.k8s.io/4006)

> ### Deprecations and removals

Reviewer 2: @graz-dev That sounds good to me! Let's do that

> ### Deprecations and removals

Author: done!

---  
layout: blog
title: "Kubernetes v1.35: TBD"
date: 2025-12-17T10:30:00-08:00
draft: true
evergreen: true
slug: kubernetes-v1-35-release
author: >
  [Kubernetes v1.35 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.35/release-team.md)
---

**Editors**: Aakanksha Bhende, Arujjwal Negi, Chad M. Crowell, Graziano Casto, Swathi Rao

Similar to previous releases, the release of Kubernetes v1.35 introduces new stable, beta, and alpha features. The consistent delivery of high-quality releases underscores the strength of our development cycle and the vibrant support from our community.

> This release consists of 59 enhancements. Of those enhancements, 16 have graduated to Stable, 19 have entered Beta, and 22 have entered Alpha.

Reviewer 2: I would keep it in the count of KEPs at least. There were changes and we processed them.

In regards to keeping the section,
I understand this feature already went GA in 1.34. Since the DRA: Structured Params
kubernetes/enhancements#4381
#52881
 is being locked in v1.35, that feels notable for users, but I'm unsure if that warrants a full section on it again in the announcement blog. I'd like to get @kubernetes/sig-node-leads advice on this

> This release consists of 59 enhancements. Of those enhancements, 16 have graduated to Stable, 19 have entered Beta, and 22 have entered Alpha.

Author: Ok updated back to 60 enhancements and 17 stable. 
Do I need to add it to the "### Graduations to stable" list?

> This release consists of 59 enhancements. Of those enhancements, 16 have graduated to Stable, 19 have entered Beta, and 22 have entered Alpha.

Reviewer 2: No, I think we can keep that the same since it's not graduated, but rather was stable and is staying there.  The math might not add up, but it's still technically correct this way

There are also some [deprecations and removals](#deprecations-and-removals) in this release; make sure to read about those.

## Release theme and logo

{{< figure src="TBD" alt="Kubernetes v1.35 logo: TBD" class="release-logo" >}}

TBD

## Spotlight on key updates

Kubernetes v1.35 is packed with new features and improvements. Here are a few select updates the Release Team would like to highlight!

### Stable: In-place update of Pod resources

Kubernetes is graduating in-place updates for Pod resources to General Availability (GA).

> This feature allows users to adjust cpu and memory resources without restarting Pods or Containers. Previously, such modifications required recreating Pods, which could disrupt workloads, particularly for stateful or batch applications. Previous Kubernetes releases already allowed you to change infrastructure resources settings (requests and limits) for existing Pods. This allows for smoother vertical scaling, improves efficiency, and can also simplify development.

Reviewer 3: ```suggestion
This feature allows users to adjust cpu and memory resources without restarting Pods or Containers. Previously, such modifications required recreating Pods, which could disrupt workloads, particularly for stateful or batch applications. Previous Kubernetes releases only allowed you to change infrastructure resources settings (requests and limits) for existing Pods. The new in-place functionality allows for smoother, nondisruptive vertical scaling, improves efficiency, and can also simplify development.
```

Author: done

This work was done as part of [KEP #1287](https://kep.k8s.io/1287) led by SIG Node.

> ### Beta: Pod certificates for workload identity and security

Reviewer 2: Hello @kubernetes/sig-auth-leads @ahmedtd! Our release comms team put this section together for https://kep.k8s.io/4317 for the v1.35 announcement blog which we're sharing with the media for an early peak, and of course, we release next week.
How does this look?

Appreciate any contribution, advice or feedback 😄

This work was done as part of [KEP #4317](https://kep.k8s.io/4317) led by SIG Auth.

### Alpha: Node declared features before scheduling

When control planes enable new features, but nodes lag behind (permitted by Kubernetes skew policy), the scheduler can place pods requiring those features onto incompatible older nodes.
The node-declaration features framework: nodes declare their supported Kubernetes features. With the new alpha feature enabled, a Node reports the features it supports, publishing this information to the control plane via a new `.status.declaredFeatures` field. Then, the `kube-scheduler`, admission controllers and third-party components can use these declarations. For example, you can enforce scheduling and API validation constraints to ensure that Pods run only on compatible nodes.

This work was done as part of [KEP #5328](https://kep.k8s.io/5328) led by SIG Node.

## Features graduating to Stable

*This is a selection of some of the improvements that are now stable following the v1.35 release.*

> ### Comparable resource version

Reviewer 2: Hello @kubernetes/sig-api-machinery-leads @michaelasp! Our release comms team put this section together for https://kep.k8s.io/5504 for the v1.35 announcement blog which will be shared with the media for an early peak by tomorrow, and of course, which we're sharing with the media for an early peak, and of course, we release next week.

Appreciate any contribution, advice or feedback 😄

> `ResourceVersion` comparability is being strengthened to give clients more than simple equality checks. Today, clients treat the resource version as an opaque string, while the apiserver uses it internally as a monotonically increasing integer. The new proposal aligns client-side semantics with apiserver behavior, allowing clients to interpret resource versions as integers and compare their ordering.

Author: done

> This enables key use cases like storage version migration, informer performance improvements, and more reliable controller behavior, all of which require knowing whether one resource version is newer than another.

Author: done

As the foundational DRA API graduated to General Availability in v1.34, v1.35 focuses on stability and ecosystem growth rather than breaking changes to this core specification. The API continues to rely on `ResourceSlice` objects for publishing device inventories and `ResourceClaims` for user requests. In v1.35, this stable foundation supports the continued evolution of advanced Alpha features such as *Device Taints* and *Consumable Capacity*, ensuring consistent and reliable resource allocation for batch and edge use cases.

This work was done as part of [KEP #4381](https://kep.k8s.io/4381) led by SIG Node.

### Reliable Pod update tracking with `Generation`

Historically, the Pod API lacked the `metadata.Generation` field found in other Kubernetes objects like Deployments. A drawback of this omission was that controllers and users had no reliable way to verify if the `kubelet` had actually processed the latest changes to a Pod's specification. This ambiguity was particularly problematic for features like [In-Place Pod Vertical Scaling](#stable-in-place-update-of-pod-resources), where knowing exactly when a resource resize request had been enacted was difficult.

> This KEP introduces standard generation tracking to the Pod API. Every time a Pod's `spec` is updated, the `metadata.Generation` sequence is incremented. Crucially, the Pod status now includes an `observedGeneration` field, which reports the generation that the `kubelet` has successfully seen and processed.

Reviewer 4: ```suggestion
This KEP introduces standard generation tracking for the Pod API. Every time a Pod's `spec` is updated, the `metadata.Generation` sequence is incremented. Crucially, the Pod status now includes an `observedGeneration` field, which reports the generation that the `kubelet` has successfully seen and processed.
```

As this feature graduates to Stable in v1.35, it is available for all workloads. This improvement provides a deterministic mechanism for users and controllers to confirm that updates, such as configuration changes or resource resizing, have been fully propagated to the running container, replacing previous guesswork with explicit status reporting.

This work was done as part of [KEP #5067](https://kep.k8s.io/5067) led by SIG Node.

### Configurable NUMA Node limit for Topology Manager

> The `TopologyManager` has historically used a hardcoded limit of 8 for the maximum number of NUMA nodes it supports to prevent state explosion during affinity calculation. A drawback of this fixed limit is that it prevents Kubernetes from fully utilizing modern high-end servers which increasingly feature CPU architectures with more than 8 NUMA nodes.

Reviewer 4: ```suggestion
The `TopologyManager` has historically used a hard-coded limit of 8 for the maximum number of NUMA nodes it supports, preventing state explosion during affinity calculation. A drawback of this fixed limit is that it prevents Kubernetes from fully utilizing modern high-end servers, which increasingly feature CPU architectures with more than 8 NUMA nodes.
```

> This KEP introduces a new `max-allowable-numa-nodes` configuration option to the Topology Manager policy. This allows cluster administrators to manually increase the limit beyond the default of 8, enabling the scheduler and Topology Manager to correctly handle resource alignment on massive SMP (Symmetric Multi-Processing) systems without requiring a code change or custom build.

Author: done

> As this feature graduates to Stable in v1.35, it is available for immediate use. Administrators managing large-scale hardware can now tune this parameter via the Kubelet configuration to match their specific hardware topology, ensuring that high-performance workloads can run optimally even on the largest available commodity servers.

Reviewer 5: Perhaps
```
As this feature graduates to Stable in v1.35, it is available for immediate use. Administrators managing large-scale hardware can now tune this parameter via the Kubelet configuration to match their specific hardware topology, ensuring that high-performance workloads can run even on the largest available commodity servers. Users should however be aware that the allocation ad admission time degrades very rapidly on machines with more than 8 NUMA nodes. 

The community is aware of the poor performance and there is a proposed enhancement in https://github.com/kubernetes/enhancements/issues/5726
```

> As this feature graduates to Stable in v1.35, it is available for immediate use. Administrators managing large-scale hardware can now tune this parameter via the Kubelet configuration to match their specific hardware topology, ensuring that high-performance workloads can run optimally even on the largest available commodity servers.

Author: done

This work was done as part of [KEP #4622](https://kep.k8s.io/4622) led by SIG Node.

## New features in Beta

*This is a selection of some of the improvements that are now beta following the v1.35 release.*

> ### Expose node topology labels via Downward API

Reviewer: @andrewsykim for text review

> ### Expose node topology labels via Downward API

Reviewer 6: This section LGTM, thanks for putting it together

Accessing node topology information, such as region and zone, from within a Pod has typically required querying the Kubernetes API server. While functional, this approach creates complexity and security risks by necessitating broad RBAC permissions or sidecar containers just to retrieve infrastructure metadata. Kubernetes v1.35 promotes the capability to expose node topology labels directly via the Downward API to beta. 

> The kubelet can now inject standard topology labels, such as `topology.kubernetes.io/zone` and `topology.kubernetes.io/region`, into Pods as environment variables or projected volume files. The primary benefit is a safer and more efficient way for workloads to be topology-aware. It allows applications to natively adapt to their availability zone or region without dependencies on the API server, strengthening security by upholding the principle of least privilege and simplifying cluster configuration. 

Reviewer 4: ```suggestion
The `kubelet` can now inject standard topology labels, such as `topology.kubernetes.io/zone` and `topology.kubernetes.io/region`, into Pods as environment variables or projected volume files. The primary benefit is a safer and more efficient way for workloads to be topology-aware. This allows applications to natively adapt to their availability zone or region without dependencies on the API server, strengthening security by upholding the principle of least privilege and simplifying cluster configuration. 
```

### Move storage version migrator in-tree

Kubernetes is improving how it handles storage migrations by introducing a built-in, streamlined mechanism that makes it easy for users to rewrite data that needs updating for maintenance activities such as schema evolution and changes in encryption at rest. Previously, the common approach was to trigger no-op updates, often by piping `kubectl get` into `kubectl replace`, but this was inefficient for large or numerous resources and especially burdensome for data-heavy objects like secrets. With the new storage migration capability, Kubernetes safely handles update conflicts and inconsistent continue tokens behind the scenes, since only the rewrite itself matters. This enhancement eliminates the need for manual scripting or deep operational knowledge, providing administrators with a simple, reliable, and automated way to ensure stored data remains current and properly maintained. 

This work was done as part of [KEP #4912](https://kep.k8s.io/4912) led by SIG API Machinery.

> ### Mutable CSINode allocatable property

Author: title changed, let us know if anything need changes in the content. 
thanks

### Mutable CSINode allocatable property

> A CSI (Container Storage Interface) driver is a Kubernetes plugin that provides a consistent way for storage systems to be exposed to containerized workloads. The `CSINode` object records details about all CSI drivers installed on a node. However, a mismatch can arise between the reported and actual attachment capacity on nodes. When volume slots are consumed after a CSI driver starts up, `kube-scheduler` may assign stateful pods to nodes without sufficient capacity, ultimately getting stuck in a `ContainerCreating` state.

Reviewer 4: ```suggestion
A CSI (Container Storage Interface) driver is a Kubernetes plugin that provides a consistent way for storage systems to be exposed to containerized workloads. The `CSINode` object records details about all CSI drivers installed on a node. However, a mismatch can arise between the reported and actual attachment capacity on nodes. When volume slots are consumed after a CSI driver starts up, the `kube-scheduler` may assign stateful pods to nodes without sufficient capacity, ultimately getting stuck in a `ContainerCreating` state.
```

> This KEP makes `CSINode.spec.drivers[*].allocatable.count` mutable so that a node’s available volume attachment capacity can be updated dynamically. It also allows CSI drivers to control how frequently the `allocatable.count` value is updated on all nodes by introducing a configurable refresh interval, defined through the `CSIDriver` object. Additionally, it automatically updates `CSINode.spec.drivers[*].allocatable.count` on detecting a failure in volume attachment due to insufficient capacity. Although this feature graduated to Beta in v1.34, it continues to be in Beta for v1.35 to allow time for feedback. 

Reviewer 4: ```suggestion
This KEP makes `CSINode.spec.drivers[*].allocatable.count` mutable so that a node’s available volume attachment capacity can be updated dynamically. It also allows CSI drivers to control how frequently the `allocatable.count` value is updated on all nodes by introducing a configurable refresh interval, defined through the `CSIDriver` object. Additionally, it automatically updates `CSINode.spec.drivers[*].allocatable.count` on detecting a failure in volume attachment due to insufficient capacity. Although this feature graduated to Beta in v1.34, it remains in Beta for v1.35 to allow time for feedback. 
```

The batching mechanism consists of two operations that can be invoked whenever needed - create and nominate. Create leads to the creation of a new set of batch information from the scheduling results of Pods that have a valid signature. Nominate uses the batching information from create to set the nominated node name from a new Pod whose signature matches the canonical Pod’s signature.

This work was done as part of [KEP #5598](https://kep.k8s.io/5598) led by SIG Scheduling.

### `maxUnavailable` for StatefulSets

A StatefulSet runs a group of Pods and maintains a sticky identity for each of those Pods. This is critical for stateful workloads requiring stable network identifiers or persistent storage. When a StatefulSet's `.spec.updateStrategy.<type>` is set to `RollingUpdate`, the StatefulSet controller will delete and recreate each Pod in the StatefulSet. It will proceed in the same order as Pod termination (from the largest ordinal to the smallest), updating each Pod one at a time.

> This KEP adds a new field to StatefulSet's `RollingUpdate` configuration settings called `maxUnavailable`, which will allow setting the maximum number of pods that can be unavailable during the update. In Kubernetes v1.35, this KEP graduates to Beta, thus being enabled by default. This new field can be set as a positive number(example: 2) or a percentage of desired Pods(example: 10%). If this field is not specified, it will default to 1 to maintain previous behavior of rolling one Pod at a time. This will allow stateful applications that can tolerate more than 1 Pod being down to finish updating faster. 

Reviewer 7: ```suggestion
This KEP adds a new field to StatefulSet's `RollingUpdate` configuration settings called `maxUnavailable`, which will allow setting the maximum number of pods that can be unavailable during the update. This setting is most effective in combination when `.spec.podManagementPolicy` is set to `Parallel`.  In Kubernetes v1.35, this KEP graduates to Beta, thus being enabled by default. This new field can be set as a positive number(example: 2) or a percentage of desired Pods(example: 10%). If this field is not specified, it will default to 1 to maintain previous behavior of rolling one Pod at a time. This will allow stateful applications that can tolerate more than 1 Pod being down to finish updating faster. 
```
/cc @helayoty

Author: suggestion implemented, thank you 👍🏻

This work was done as part of [KEP #961](https://kep.k8s.io/961) led by SIG Apps.

### Separate `kubectl` user preferences from cluster configs

The `kubeconfig` file currently has a `Preferences` field for user preferences which is currently underutilized. A drawback of using this existing field is that it does not separate server configuration and user preferences. 

This KEP introduces a new opt-in `kuberc` file as a way to separate server configurations and cluster credentials from user preferences without disrupting already running CI pipelines with unexpected outputs. 

> As this feature graduates to Beta in v1.35, it has been enabled by default. However, users can disable it by setting the `KUBECTL_KUBERC` environment variable to false or `KUBERC` environment variable to off. Users can either create `kuberc` files manually or type in `kubectl kuberc` into the terminal to get the list of supported subcommands. Furthermore, `kubectl` will be equipped with a mechanism which will allow it to read all past versions of the `kuberc` file, and pick the latest known one. This will ensure that users can continue using the version of `kuberc` they started with, unless they are interested in newer features available only in newer releases.

Author: done! thank you :)

This work was done as part of [KEP #3104](https://kep.k8s.io/3104) led by SIG CLI.

### KYAML

YAML is a human-readable format of data serialization. In Kubernetes, YAML files are used to define and configure resources, such as pods, services and deployments. However, complex YAML is difficult to read. YAML's significant whitespace requires careful attention to indentation and nesting, while its optional string-quoting can lead to unexpected type coercion(see: The Norway Bug). While JSON is an alternative, it lacks support for comments and has strict requirements for trailing commas and quoted keys.

> KYAML is a safer and less ambiguous dialect of YAML designed specifically for Kubernetes. Introduced as an opt-in alpha feature in v1.34, this feature graduated to Beta in Kubernetes v1.35 and has been enabled by default. It can be disabled by setting the environment variable `KUBECTL_KYAML=false`. 

Reviewer 8: (drive-by comment: I haven't paid attention to all of the discussion around KYAML and how we're talking about it, so feel free to ignore me...)

"dialect" could be misinterpreted as implying not 100% compatible. Maybe something like:

> KYAML is a set of best practices for formatting `kubectl`'s YAML output, adding back some of the optional punctuation that is normally omitted in YAML, to make it easier to copy/paste/merge bits of YAML without having to worry about indentation and other YAML pitfalls.

and give an example of "plain" YAML vs KYAML?

Author: Updated changing "dialect" with "subset" is it enough?
PTAL

This work was done as part of [KEP #127](https://kep.k8s.io/127) led by SIG Node.

### VolumeSource: OCI artifact and/or image

When creating a Pod, you often need to provide data, binaries, or configuration files for your containers. This meant including the content into the main container image or using a custom init container to download and unpack files into an `emptyDir`. Both these approaches are still valid. Kubernetes v1.31 added support for the `image` volume type allowing Pods to declaratively pull and unpack OCI container image artifacts into a volume. This lets you package and deliver data-only artifacts such as configs, binaries, or machine learning models using standard OCI registry tools. 

> With this feature, you can fully separate your data from your container image and remove the need for extra init containers or startup scripts. The image volume type has been in beta since v1.33 and will be enabled by default in v1.35.

Reviewer: please mention the Containerd version that has full support for the feature.

This KEP introduces a mechanism where the `kubelet` enforces credential verification for cached images. Before allowing a Pod to use a locally cached image, the `kubelet` checks if the Pod has the valid credentials to pull it. This ensures that only authorized workloads can use private images, regardless of whether they are already present on the node, significantly hardening the security posture for shared clusters.

As this feature graduates to Beta in v1.35, it has been enabled by default. However, users can disable it by setting the `KubeletEnsureSecretPulledImages` feature gate to false. The feature works by tracking the credentials used to pull images and verifying that subsequent requests for the same image provide matching or valid credentials before launching the container.

This work was done as part of [KEP #2535](https://kep.k8s.io/2535) led by SIG Node.

> ### Fine-grained Container restart rules

Reviewer: @yuanwang04 please review

The `restartPolicy` field is currently defined at the Pod level, applying the same behavior to all containers within the Pod. A drawback of this global setting is that it lacks granularity for complex workloads, such as AI/ML training jobs, where `restartPolicy: Never` is often required for the Pod but individual containers might benefit from in-place restarts upon encountering specific retriable errors.

This KEP introduces `restartPolicy` and `restartPolicyRules` fields to the container API, allowing users to define restart strategies for individual containers independent of the Pod's overall policy. This feature enables granular control, such as restarting a container only when it exits with a specific error code, thereby avoiding expensive and time-consuming Pod re-scheduling for transient failures.

> As this feature graduates to Beta, it is enabled by default. Users can utilize the new `restartPolicyRules` within the container `spec` to specify exit codes that should trigger a restart, effectively overriding the Pod-level directive for those specific conditions. This ensures that users can optimize resource utilization and recovery times for long-running workloads without disrupting existing Pod lifecycle management logic.

Reviewer 9: Can we add this does not apply to ephemeral containers and sidecar containers / or only applies to regular and init containers?

```suggestion
As this feature graduates to Beta, it is enabled by default. Users can utilize the new `restartPolicyRules` within the regular container's and init container's `spec` to specify exit codes that should trigger a restart, effectively overriding the Pod-level directive for those specific conditions. This ensures that users can optimize resource utilization and recovery times for long-running workloads without disrupting existing Pod lifecycle management logic.
```

### CSI driver opt-in for service account tokens via secrets field

Providing ServiceAccount tokens to Container Storage Interface (CSI) drivers has traditionally relied on injecting them into the `volume_context` field. This approach presents a significant security risk because `volume_context` is intended for non-sensitive configuration data and is frequently logged in plain text by drivers and debugging tools, potentially leaking credentials. 

Kubernetes v1.35 introduces an opt-in mechanism for CSI drivers to receive ServiceAccount tokens via the dedicated secrets field in the NodePublishVolume request. Drivers can now enable this behavior by setting the `serviceAccountTokenInSecrets` field to true in their CSIDriver object, instructing the `kubelet` to populate the token securely. 

The primary benefit is the prevention of accidental credential exposure in logs and error messages. This change ensures that sensitive workload identities are handled via the appropriate secure channels, aligning with best practices for secret management while maintaining backward compatibility for existing drivers. 

> This work was done as part of [KEP #5538](https://kep.k8s.io/5538) led by SIG Storage.

Reviewer 10: This work was done by SIG Auth in cooperation with SIG Storage.

### Extended toleration operators for threshold-based placement

Managing workload prioritization in Kubernetes has typically depended on PriorityClass, which orders Pods but does not account for waiting time. This static approach can lead to resource starvation for lower-priority workloads where they are perpetually preempted by higher-priority tasks making it difficult to guarantee strict Service Level Agreements (SLAs).

Kubernetes v1.35 introduces SLA-aware scheduling capabilities within the Scheduling Framework. This enhancement updates the `QueueSort` plugin to optionally consider the waiting time of Pods, effectively elevating the priority of long-pending workloads to ensure they eventually get scheduled.

The primary benefit is a fairer and more deterministic scheduling lifecycle. By mitigating resource starvation, operators can enforce stricter time-based SLAs and ensure that lower-priority batch jobs or background tasks make steady progress even in highly utilized clusters.

This work was done as part of [KEP #5471](https://kep.k8s.io/5471) led by SIG Scheduling.

### Consider terminating Pods in Deployments

> Updating or scaling Deployments has historically relied on an eager strategy that creates replacement Pods immediately, often ignoring those that are still terminating. This behavior can lead to resource contention and scheduling deadlocks, particularly in clusters with limited capacity, as terminating Pods continue to occupy quota while new replicas compete for the same resources.

Reviewer 7: This entire paragraph is not correct. As part of the overall work towards exposing that functionality you're writing about we ONLY exposed the information about terminating replicas in the deployment. That functionality is being promoted to beta, the rest of the work under [KEP-3973](https://kep.k8s.io/3973) umbrella is still WIP.

Reviewer 8: The title of the section is also confusing. I parsed it as "You should consider terminating the Pods in your Deployments".

Reviewer 2: Scratch that, we see @dipesh-rawat's comment below
> I remember that this https://github.com/kubernetes/enhancements/issues/3973 could only merge some of its changes before the code freeze. Only DeploymentReplicaSetTerminatingReplicas feature gate is moved to beta for this release, although the main feature DeploymentPodReplacementPolicy is still in alpha.

> There is some additional context in latest comment post code freeze https://github.com/kubernetes/enhancements/issues/3973#issuecomment-3543463447.

> @atiratree Please feel free to jump in and help us move this forward.

Kubernetes v1.35 introduces the `podReplacementPolicy` field for Deployments, enabling users to enforce a TerminationComplete strategy. This new mechanism ensures that the controller waits for terminating Pods to fully exit before creating replacements, while also providing visibility into terminating replicas via the Deployment status.

> The primary benefit is safer and more predictable application rollouts. By serializing the termination and creation steps, it prevents transient resource overcommitment, making it easier to manage workloads with complex shutdown sequences in high-density environments.

Reviewer 7: Maybe something like:

```
Kubernetes v1.35 introduces the `terminatingReplicas` field for Deployments, 
informing users about the total number of terminating pods targeted by this 
deployment. This field is part of an ongoing major initiative which plans to 
improve the pod replacement policy inside a deployment. 
```

/cc @atiratree

> The primary benefit is safer and more predictable application rollouts. By serializing the termination and creation steps, it prevents transient resource overcommitment, making it easier to manage workloads with complex shutdown sequences in high-density environments.

Reviewer 7: Also, since this field is beta, it might need to move to previous section, right?

> The primary benefit is safer and more predictable application rollouts. By serializing the termination and creation steps, it prevents transient resource overcommitment, making it easier to manage workloads with complex shutdown sequences in high-density environments.

Reviewer 11: I remember that this [KEP-3973](https://github.com/kubernetes/enhancements/issues/3973) could only merge some of its changes before the code freeze. Only `DeploymentReplicaSetTerminatingReplicas` feature gate is moved to beta for this release, although the main feature `DeploymentPodReplacementPolicy` is still in alpha. 

There is some additional context in latest comment post code freeze https://github.com/kubernetes/enhancements/issues/3973#issuecomment-3543463447. 

@atiratree Please feel free to jump in and help us move this forward.

During the v1.35 release cycle, which spanned 14 weeks from 15th September 2025 to 17th December 2025, Kubernetes received contributions from as many as 85 different companies and 419 individuals. In the wider cloud native ecosystem, the figure goes up to 281 companies, counting 1769 total contributors.

Note that "contribution" counts when someone makes a commit, code review, comment, creates an issue or PR, reviews a PR (including blogs and documentation) or comments on issues and PRs.  
If you are interested in contributing, visit [Getting Started](https://www.kubernetes.dev/docs/guide/#getting-started) on our contributor website.

Source for this data:

* [Companies contributing to Kubernetes](https://k8s.devstats.cncf.io/d/11/companies-contributing-in-repository-groups?orgId=1&from=1757890800000&to=1765929599000&var-period=d28&var-repogroup_name=Kubernetes&var-repo_name=kubernetes%2Fkubernetes)  
* [Overall ecosystem contributions](https://k8s.devstats.cncf.io/d/11/companies-contributing-in-repository-groups?orgId=1&from=1757890800000&to=1765929599000&var-period=d28&var-repogroup_name=All&var-repo_name=kubernetes%2Fkubernetes)

> ## Events update

Reviewer 2: It would be more ideal to have links, but if there is no page, just listing the event without a link seems like the best we can do.

> ## Events update

Reviewer 2: x-post: https://kubernetes.slack.com/archives/C1TU9EB9S/p1765322100970639

also @graz-dev,
@dipesh-rawat recommends that ultimately, we remove any of these KCDs that don't have links. I can see this go either way, so figure we could discuss it

---  
layout: blog
title: "Kubernetes v1.35: TBD"
date: 2025-12-17T10:30:00-08:00
draft: true
evergreen: true
slug: kubernetes-v1-35-release
author: >
  [Kubernetes v1.35 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.35/release-team.md)
---

**Editors**: Aakanksha Bhende, Arujjwal Negi, Chad M. Crowell, Graziano Casto, Swathi Rao

Similar to previous releases, the release of Kubernetes v1.35 introduces new stable, beta, and alpha features. The consistent delivery of high-quality releases underscores the strength of our development cycle and the vibrant support from our community.

This release consists of 60 enhancements. Of those enhancements, 17 are Stable, 19 Beta, and 22 Alpha.

There are also some [deprecations and removals](#deprecations-and-removals) in this release; make sure to read about those.

## Release theme and logo

{{< figure src="TBD" alt="Kubernetes v1.35 logo: TBD" class="release-logo" >}}

> TBD

Reviewer 7: I believe this can be updated now that we have https://github.com/kubernetes/sig-release/tree/master/releases/release-1.35/logo

The goal is to allow clean delegation of Job synchronization to another controller. It does not aim to pass custom parameters to that controller or modify CronJob concurrency policies.

This work was done as part of [KEP #4368](https://kep.k8s.io/4368) led by SIG Apps.

### Reliable Pod update tracking with `Generation`

Historically, the Pod API lacked the `metadata.Generation` field found in other Kubernetes objects such as Deployments. A drawback of this omission was that controllers and users had no reliable way to verify if the `kubelet` had actually processed the latest changes to a Pod's specification. This ambiguity was particularly problematic for features like [In-Place Pod Vertical Scaling](#stable-in-place-update-of-pod-resources), where knowing exactly when a resource resize request had been enacted was difficult.

> This KEP introduces standard generation tracking for the Pod API. Every time a Pod's `spec` is updated, the `metadata.Generation` sequence is incremented. Crucially, the Pod status now includes a top-level `observedGeneration` field, which reports the generation that the `kubelet` has successfully seen and processed. Pod conditions also each contain their own individual `observedGeneration` field that can

Author: done

This work was done as part of [KEP #4622](https://kep.k8s.io/4622) led by SIG Node.

## New features in Beta

*This is a selection of some of the improvements that are now beta following the v1.35 release.*

### Expose node topology labels via Downward API

Accessing node topology information, such as region and zone, from within a Pod has typically required querying the Kubernetes API server. While functional, this approach creates complexity and security risks by necessitating broad RBAC permissions or sidecar containers just to retrieve infrastructure metadata. Kubernetes v1.35 promotes the capability to expose node topology labels directly via the Downward API to beta. 

The `kubelet` can now inject standard topology labels, such as `topology.kubernetes.io/zone` and `topology.kubernetes.io/region`, into Pods as environment variables or projected volume files. The primary benefit is a safer and more efficient way for workloads to be topology-aware. This allows applications to natively adapt to their availability zone or region without dependencies on the API server, strengthening security by upholding the principle of least privilege and simplifying cluster configuration. 

> Note: this KEP injects topology labels to every Pod so they can be configured for Downward API. With 1.35 upgrade, you will see a lot of new labels on each Pod, and this is a part of a design.

Author: done!

This KEP adds a new field to StatefulSet's `RollingUpdate` configuration settings called `maxUnavailable`, which will allow setting the maximum number of pods that can be unavailable during the update. This setting is most effective in combination when `.spec.podManagementPolicy` is set to `Parallel`.  In Kubernetes v1.35, this KEP graduates to Beta, thus being enabled by default. This new field can be set as a positive number(example: 2) or a percentage of desired Pods(example: 10%). If this field is not specified, it will default to 1 to maintain previous behavior of rolling one Pod at a time. This will allow stateful applications that can tolerate more than 1 Pod being down to finish updating faster. 

This work was done as part of [KEP #961](https://kep.k8s.io/961) led by SIG Apps.

### Configure credential plugin policy in `kuberc`

Optional [`kuberc` file](/docs/reference/kubectl/kuberc/) is a way to separate server configurations and cluster credentials from user preferences without disrupting already running CI pipelines with unexpected outputs.

> As part of the v1.35 release `kuberc` gains additional an functionality which allows users to configure credential plugin policy. This change introduces two fields `credentialPluginPolicy`, which allows or denies all plugins, and allows specifying a list of allowed plugins using `credentialPluginAllowlist`.

Author: done, thanks

This work was done as part of [KEP #127](https://kep.k8s.io/127) led by SIG Node.

### VolumeSource: OCI artifact and/or image

When creating a Pod, you often need to provide data, binaries, or configuration files for your containers. This meant including the content into the main container image or using a custom init container to download and unpack files into an `emptyDir`. Both these approaches are still valid. Kubernetes v1.31 added support for the `image` volume type allowing Pods to declaratively pull and unpack OCI container image artifacts into a volume. This lets you package and deliver data-only artifacts such as configs, binaries, or machine learning models using standard OCI registry tools. 

> With this feature, you can fully separate your data from your container image and remove the need for extra init containers or startup scripts. The image volume type has been in beta since v1.33 and will be enabled by default in v1.35.

Reviewer 12: "will be"?

The [core functionality](https://kep.k8s.io/4381) was graduated to stable in 1.34, with the ability to turn it off. In 1.35 it is always enabled. Several alpha features were improved and are ready for a final round of feedback from users before promoting them to beta in 1.36.

#### [Extended Resource Requests via DRA](https://kep.k8s.io/5004)

Several functional gaps compared to Extended Resource requests via Device Plugins were addressed, for example scoring and reuse of devices in init containers.

#### [Device Taints and Tolerations](https://kep.k8s.io/5055)

The new "None" effect can be used to report a problem without immediately affecting scheduling or running pod. DeviceTaintRule now provides status information about an on-going eviction. The "None" effect can be used for a "dry run" before actually evicting pods:
- Create DeviceTaintRule with "effect: None".
- Check the status to see how many pods would be evicted.
- Replace "effect: None" with "effect: NoExecute".

#### [Partitionable Devices](https://kep.k8s.io/4815)

Devices belonging to the same partitionable devices may now be defined in different ResourceSlices.

> 

Reviewer 12: If there are (will be) docs, link to them.

#### [Consumable Capacity](https://kep.k8s.io/5075), [Device Binding Conditions](https://kep.k8s.io/5007)

Several bugs were fixed and/or more tests added.

> 

Reviewer 12: If there are (will be) docs, link to them.

### Deployment status: count of terminating replicas

Historically, the Deployment status provided details on available and updated replicas but lacked explicit visibility into Pods that were in the process of shutting down. A drawback of this omission was that users and controllers could not easily distinguish between a stable Deployment and one that still had Pods executing cleanup tasks or adhering to long grace periods.

Kubernetes v1.35 promotes the `terminatingReplicas` field within the Deployment status to Beta. This field provides a count of Pods that have a deletion timestamp set but have not yet been removed from the system. This feature is a foundational step in a larger initiative to improve how Deployments handle Pod replacement, laying the groundwork for future policies regarding when to create new Pods during a rollout.

The primary benefit is improved observability for lifecycle management tools and operators. By exposing the number of terminating Pods, external systems can now make more informed decisions such as waiting for a complete shutdown before proceeding with subsequent tasks without needing to manually query and filter individual Pod lists.

> This work was done as part of [KEP #3973](https://kep.k8s.io/3973) led by SIG Apps.

Reviewer 7: This looks great now, thank you! But I believe it should move to beta section, right?

This work was done as part of [KEP #5328](https://kep.k8s.io/5328) led by SIG Node.

## Features graduating to Stable

*This is a selection of some of the improvements that are now stable following the v1.35 release.*

### Comparable ResourceVersion semantics

The `ResourceVersion` comparability has been strengthened to give clients more than simple equality checks. With Kubernetes v1.34 or before, clients treat the resource version as an opaque string, while the API server uses it internally as a monotonically increasing integer. In v1.35, this proposal aligns client-side semantics with API server behavior, allowing clients to interpret resource versions as integers and compare their ordering.
This enables key use cases such as storage version migration, informer performance improvements, and more reliable controller behavior, all of which require knowing whether one resource version is newer than another.

> The KEP introduces a comparison utility for clients and adds conformance tests to ensure clusters meet the new guarantees. It does not attempt to limit resource version size or define additional structure beyond comparability.

Reviewer 12: **It's not a graduation to stable.**

Graduations mean that a feature was previously alpha or beta and is now stable; that's not true here.
This is a semantic change that is instantly stable when v1.35 is released.

I'd move it elsewhere; also, here's how I'd write it up:
```markdown
## Comparable resource version semantics

Kubernetes v1.35 changes the way that clients are allowed to interpret
[resource versions](/docs/reference/using-api/api-concepts/#resource-versions).

Before v1.35, the only supported comparison that clients could make was to
check for string equality: if two resource versions were equal, they were the same.
Clients could also provide a resource version to the API server and ask the
control plane to do internal comparisons, such as streaming all events since
a particular resource version.

In v1.35, all in-tree resource versions meet a new stricter definition: the
values are a special form of decimal number. And, because they can be compared,
clients can do their own operations to compare two different resource versions.
For example, this means that a client reconnecting after a crash can detect
when it has lost updates, as distinct from the case where there has been an 
update but no lost changes in the meantime.

This change in semantics enables other important use cases such as [storage version migration](/docs/tasks/manage-kubernetes-objects/storage-version-migration/),
performance improvements to _informers_ (a client helper concept), and controller reliability. All of those cases require knowing whether one resource version is newer than another.
```

(notice how I don't use the term "KEP"; I don't think our readers need to know that the enhancement has been proposed. They can tell in any case that is has actually happened).

Author: done.
@drewhagen where should we move it?

Reviewer 2: I think let's add in that "Other Notable Changes" section actually, and move it under that. I agree with others that it is worth calling out.

This work was done as part of [KEP #3015](https://kep.k8s.io/3015) led by SIG Network.

### Job API managed-by mechanism

This KEP introduces support for the `managedBy` field allowing an external controller to handle Job status synchronization. The feature is driven by [MultiKueue](https://github.com/kubernetes-sigs/kueue/tree/main/keps/693-multikueue), a multi-cluster Job dispatching system where a Job created in a management cluster is mirrored and executed in a worker cluster, with status propagated back. To enable this workflow, the built-in Job controller must not act on a particular Job resource so that the Kueue controller can manage status updates instead.

The goal is to allow clean delegation of Job synchronization to another controller. It does not aim to pass custom parameters to that controller or modify CronJob concurrency policies.

This work was done as part of [KEP #4368](https://kep.k8s.io/4368) led by SIG Apps.

> ### Reliable Pod update tracking with `Generation`

Author: done

A drawback of this omission was that controllers and users had no reliable way to verify if the `kubelet` had actually processed the latest changes to a Pod's specification. This ambiguity was particularly problematic for features like [In-Place Pod Vertical Scaling](#stable-in-place-update-of-pod-resources), where it was difficult to know exactly when a resource resize request had been enacted.

Kubernetes v1.33 added `.metadata.generation` fields for Pods, as an alpha feature. That field is now stable in the v1.35 Pod API, which means that every time a Pod's `spec` is updated, the `.metadata.generation` value is incremented. As part of this improvement, the Pod API also gained a `.status.observedGeneration` field, which reports the generation that the `kubelet` has successfully seen and processed. Pod conditions also each contain their own individual `observedGeneration` field, that clients can report and / or observe.

Because this feature has graduated to stable in v1.35, it is available for all workloads.

This work was done as part of [KEP #5067](https://kep.k8s.io/5067) led by SIG Node.

> ### Configurable NUMA Node limit for Topology Manager

Author: done

> The `TopologyManager` has historically used a hard-coded limit of 8 for the maximum number of NUMA nodes it supports, preventing state explosion during affinity calculation. A drawback of this fixed limit is that it prevents Kubernetes from fully utilizing modern high-end servers, which increasingly feature CPU architectures with more than 8 NUMA nodes.

Reviewer 12: Maybe this:
```suggestion
The [topology manager](/docs/concepts/policy/node-resource-managers/) historically used a hard-coded limit of 8 for the maximum number of NUMA nodes it can support,
preventing state explosion during affinity calculation.
(There's an important detail here; a _NUMA node_ is not the same as a Node in the Kubernetes API).
This limit on the number of NUME nodes prevented Kubernetes from fully utilizing modern high-end servers,
which increasingly feature CPU architectures with more than 8 NUMA nodes.
```

> This KEP introduces a new `max-allowable-numa-nodes` configuration option to the Topology Manager policy. This allows cluster administrators to manually increase the limit beyond the default of 8. Prior to this enhancement, running the `kubelet` on a machine with more than 8 NUMA nodes and configuring the Topology Manager with a policy different from None would have caused the `kubelet` to fail to start. With this configuration option, cluster administrators can use these machines with more than 8 NUMA nodes.

Author: done

As this feature graduates to Stable in v1.35, it is available for immediate use, you can read more in the [official documentation](/docs/tasks/administer-cluster/topology-manager/#policy-option-max-allowable-numa-nodes).

> The community is aware of the poor performance and there is a proposed enhancement in [KEP #5726](https://kep.k8s.io/5726).

Author: done

The `kubelet` can now inject standard topology labels, such as `topology.kubernetes.io/zone` and `topology.kubernetes.io/region`, into Pods as environment variables or projected volume files. The primary benefit is a safer and more efficient way for workloads to be topology-aware. This allows applications to natively adapt to their availability zone or region without dependencies on the API server, strengthening security by upholding the principle of least privilege and simplifying cluster configuration. 

**Note:** Kubernetes now injects available topology labels to every Pod so that they can be used as inputs to the [downward API](/docs/concepts/workloads/pods/downward-api/). With the v1.35 upgrade, most cluster administrators will see several new labels added to each Pod; this is expected as part of the design.

This work was done as part of [KEP #4742](https://kep.k8s.io/4742) led by SIG Node.

> ### Move storage version migrator in-tree

Reviewer 12: (nit; does not block a merge)
```suggestion
### In-tree support for storage version migration
```
?

We may want to explain what we mean by "in tree"

Kubernetes v1.24 added a new **alpha** field to a StatefulSet's `rollingUpdate` configuration settings, called `maxUnavailable`. That field wasn't part of the Kubernetes API unless your cluster administrator explicitly opted in.
In Kubernetes v1.35 that field is beta and is available by default. You can use it to define the maximum number of pods that can be unavailable during an update. This setting is most effective in combination with `.spec.podManagementPolicy` set to Parallel.  You can set `maxUnavailable` as either a positive number (example: 2) or a percentage of the desired number of Pods (example: 10%). If this field is not specified, it will default to 1, to maintain the previous behavior of only updating one Pod at a time. This improvement allows stateful applications (that can tolerate more than one Pod being down) to finish updating faster.

This work was done as part of [KEP #961](https://kep.k8s.io/961) led by SIG Apps.

> ### Configure credential plugin policy in `kuberc`

Reviewer 12: (nit; does not block a merge)
```suggestion
### Configurable credential plugin policy in `kuberc`
```

### KYAML

YAML is a human-readable format of data serialization. In Kubernetes, YAML files are used to define and configure resources, such as Pods, Services, and Deployments. However, complex YAML is difficult to read. YAML's significant whitespace requires careful attention to indentation and nesting, while its optional string-quoting can lead to unexpected type coercion (see: The Norway Bug). While JSON is an alternative, it lacks support for comments and has strict requirements for trailing commas and quoted keys.

KYAML is a safer and less ambiguous subset of YAML designed specifically for Kubernetes. Introduced as an opt-in alpha feature in v1.34, this feature graduated to Beta in Kubernetes v1.35 and has been enabled by default. It can be disabled by setting the environment variable `KUBECTL_KYAML=false`. 

KYAML addresses challenges pertaining to both YAML and JSON. All KYAML files are also valid YAML files. This means you can write KYAML and pass it as an input to any version of kubectl. This also means that you don’t need to write in strict KYAML for the input to be parsed.

This work was done as part of [KEP #5295](https://kep.k8s.io/5295) led by SIG CLI.

> ### Configurable tolerance for Horizontal Pod Autoscalers

Reviewer 12: (nit; does not block a merge)
```suggestion
### Configurable tolerance for HorizontalPodAutoscalers
```

Author: done

This work was done as part of [KEP #127](https://kep.k8s.io/127) led by SIG Node.

### VolumeSource: OCI artifact and/or image

When creating a Pod, you often need to provide data, binaries, or configuration files for your containers. This meant including the content into the main container image or using a custom init container to download and unpack files into an `emptyDir`. Both these approaches are still valid. Kubernetes v1.31 added support for the `image` volume type allowing Pods to declaratively pull and unpack OCI container image artifacts into a volume. This lets you package and deliver data-only artifacts such as configs, binaries, or machine learning models using standard OCI registry tools. 

With this feature, you can fully separate your data from your container image and remove the need for extra init containers or startup scripts. The image volume type has been in Beta since v1.33 and is enabled by default in v1.35. Please note that using this feature requires a compatible container runtime, such as containerd v2.1 or later.

This work was done as part of [KEP #4639](https://kep.k8s.io/4639) led by SIG Node.

> ### Ensure `kubelet` verifies credentials for cached images

Reviewer 12: (nit; does not block a merge)
```suggestion
### Enforced `kubelet` credential verification for cached images
```

Author: done

Kubernetes v1.35 introduces the `/flagz` endpoint, a built-in mechanism for components to expose their command line options. This feature allows authorized users to inspect the effective configuration via a standardized HTTP interface, supporting both plain text and structured (JSON) formats. The primary benefit is enhanced observability and secure configuration auditing. It provides a direct method to validate operational parameters without exposing the underlying infrastructure, simplifying debugging workflows and ensuring compliance with intended states.

This work was done as part of [KEP #4828](https://kep.k8s.io/4828) led by SIG Instrumentation.

### Statusz for Kubernetes components

Troubleshooting Kubernetes components like the `kube-apiserver` or `kubelet` typically involves sifting through extensive logs or relying on external monitoring tools to gather the basic operational state. This fragmented approach often delays diagnosis, as there is no standardized, in-process mechanism to quickly retrieve a snapshot of a component's internal health and status.

> Kubernetes v1.35 introduces the `/statusz` endpoint across core components, offering a unified interface for real-time status introspection. This feature provides authorized users with immediate access to structured status data such as version information and health indicators, directly from the running binary. The primary benefit is a consistent and accessible layer for debugging and observability. By standardizing how components report their state, it simplifies the initial triage of performance issues and version mismatches, reducing the time to resolution without requiring complex external setups.

Author: done!

This work was done as part of [KEP #4827](https://kep.k8s.io/4827) led by SIG Instrumentation.

### CCM: watch-based route controller reconciliation using informers

Managing network routes within cloud environments has traditionally relied on the Cloud Controller Manager (CCM) periodically polling the cloud provider's API to verify and update route tables. This fixed-interval reconciliation approach can be inefficient, often generating a high volume of unnecessary API calls and introducing latency between a node state change and the corresponding route update.

> Kubernetes v1.35 introduces a watch-based reconciliation strategy for the route controller. Instead of relying on a timer, the controller now utilizes informers to watch for specific Node events, such as additions, deletions, or relevant field updates and triggers route synchronization only when a change actually occurs.

Author: done

This work was done as part of [KEP #5440](https://kep.k8s.io/5440) led by SIG Apps.

## Graduations, deprecations, and removals in v1.35

### Graduations to stable

This lists all the features that graduated to stable (also known as *general availability*). For a full list of updates including new features and graduations from alpha to beta, see the release notes.

This release includes a total of 16 enhancements promoted to stable:

* [Add CPUManager policy option to restrict reservedSystemCPUs to system daemons and interrupt processing](https://kep.k8s.io/4540)
* [Pod Generation](https://kep.k8s.io/5067)
* [Invariant Testing](https://kep.k8s.io/5468)
* [In-Place Update of Pod Resources](https://kep.k8s.io/1287)
* [Fine-grained SupplementalGroups control](https://kep.k8s.io/3619)
* [Add support for a drop-in kubelet configuration directory](https://kep.k8s.io/3983)
* [Remove gogo protobuf dependency for Kubernetes API types](https://kep.k8s.io/5589)
* [kubelet image GC after a maximum age](https://kep.k8s.io/4210)
* [Kubelet limit of Parallel Image Pulls](https://kep.k8s.io/3673)

> * [Comparable Resource Version](https://kep.k8s.io/5504)

Reviewer 2: I think let's add in that "Other Notable Changes" section actually, and move it under that. I agree with @lmktfy and SIG API Machinery's own recommendation that it is worth calling out.

> * [Comparable Resource Version](https://kep.k8s.io/5504)

Author: done

At its center sits the Kubernetes wheel wrapped around the Earth, grounded by the resilient maintainers, contributors and users who keep showing up. Between day jobs, life changes, and steady open-source stewardship, they prune old APIs, graft new features and keep one of the world’s largest open source projects healthy.

Three squirrels guard the tree: a wizard holding the LGTM scroll for reviewers, a warrior with an axe and Kubernetes shield for the release crews who cut new branches, and a rogue with a lantern for the triagers who bring light to dark issue queues.

Together, they stand in for a much larger adventuring party. Kubernetes v1.35 adds another growth ring to the World Tree, a fresh cut shaped by many hands, many paths and a community whose branches reach higher as its roots grow deeper.

## Spotlight on key updates

Kubernetes v1.35 is packed with new features and improvements. Here are a few select updates the Release Team would like to highlight!

### Stable: In-place update of Pod resources

Kubernetes has graduated in-place updates for Pod resources to General Availability (GA).

> This feature allows users to adjust CPU and memory resources without restarting Pods or Containers. Previously, such modifications required recreating Pods, which could disrupt workloads, particularly for stateful or batch applications. Earlier Kubernetes releases allowed you to change only infrastructure resources settings (requests and limits) for existing Pods. The new in-place functionality allows for smoother, nondisruptive vertical scaling, improves efficiency, and can also simplify development.

Reviewer 4: ```suggestion
This feature allows users to adjust CPU and memory resources without restarting Pods or Containers. Previously, such modifications required recreating Pods, which could disrupt workloads, particularly for stateful or batch applications. Earlier Kubernetes releases allowed you to change only infrastructure resource settings (requests and limits) for existing Pods. The new in-place functionality allows for smoother, nondisruptive vertical scaling, improves efficiency, and can also simplify development.
```

For the Kubernetes v1.35 release, the cloud-controller-manager library introduces a watch-based reconciliation strategy for the route controller. Instead of relying on a timer, the controller now utilizes informers to watch for specific Node events, such as additions, deletions, or relevant field updates and triggers route synchronization only when a change actually occurs.

The primary benefit is a significant reduction in cloud provider API usage, which lowers the risk of hitting rate limits and reduces operational overhead. Additionally, this event-driven model improves the responsiveness of the cluster's networking layer by ensuring that route tables are updated immediately following changes in cluster topology.

This work was done as part of [KEP #5237](https://kep.k8s.io/5237) led by SIG Cloud Provider.

### Extended toleration operators for threshold-based placement

Kubernetes v1.35 introduces SLA-aware scheduling by enabling workloads to express reliability requirements. The feature adds numeric comparison operators to tolerations, allowing pods to match or avoid nodes based on SLA-oriented taints such as service guarantees or fault-domain quality.

> The primary benefit is enhancing the scheduler with more precise placement. As critical workloads can demand higher-SLA nodes, while lower priority workloads can opt into lower SLA ones. This improves utilization and reduces cost without compromising reliability.

Reviewer 4: ```suggestion
The primary benefit is enhancing the scheduler with more precise placement. Critical workloads can demand higher-SLA nodes, while lower priority workloads can opt into lower SLA ones. This improves utilization and reduces cost without compromising reliability.
```

* [kubelet image GC after a maximum age](https://kep.k8s.io/4210)
* [Kubelet limit of Parallel Image Pulls](https://kep.k8s.io/3673)
* [Add a TopologyManager policy option for MaxAllowableNUMANodes](https://kep.k8s.io/4622)
* [Include kubectl command metadata in http request headers](https://kep.k8s.io/859)
* [PreferSameNode Traffic Distribution (formerly PreferLocal traffic policy / Node-level topology)](https://kep.k8s.io/3015)
* [Job API managed-by mechanism](https://kep.k8s.io/4368)
* [Transition from SPDY to WebSockets](https://kep.k8s.io/4006)

### Deprecations, removals and community updates

As Kubernetes develops and matures, features may be deprecated, removed, or replaced with better
ones to improve the project's overall health. See the Kubernetes
[deprecation and removal policy](/docs/reference/using-api/deprecation-policy/) for more details on this process. Kubernetes v1.35 includes a couple of deprecations.

#### Ingress NGINX retirement

For years, the Ingress NGINX controller has been a popular choice for routing traffic into Kubernetes clusters. It was flexible, widely adopted, and served as the standard entry point for countless applications. 

> However, maintaining the project has become unsustainable. With a severe shortage of maintainers and mounting technical debt the community recently made the difficult decision to retire it. This isn't strictly part of the v1.35 release, but it's such an important change that we wanted to highlight it here.

Reviewer 4: ```suggestion
However, maintaining the project has become unsustainable. With a severe shortage of maintainers and mounting technical debt, the community recently made the difficult decision to retire it. This isn't strictly part of the v1.35 release, but it's such an important change that we wanted to highlight it here.
```

Because of this maintenance burden, Kubernetes v1.35 deprecates `ipvs` mode. Although the mode remains available in this release, `kube-proxy` will now emit a warning on startup when configured to use it. The goal is to streamline the codebase and focus on modern standards. For Linux nodes, you should begin transitioning to [`nftables`](/docs/reference/networking/virtual-ips/#proxy-mode-nftables), which is now the recommended replacement.

You can find more in [KEP-5495: Deprecate ipvs mode in kube-proxy](https://kep.k8s.io/5495)

#### Final call for containerd v1.X

While Kubernetes v1.35 still supports containerd 1.7 and other LTS releases, this is the final version with such support. The SIG Node community has designated v1.35 as the last release to support the containerd v1.X series.

This serves as an important reminder: before upgrading to the next Kubernetes version, you must switch to containerd 2.0 or later. To help identify which nodes need attention, you can monitor the `kubelet_cri_losing_support` metric within your cluster.

> You can find more in the [official blog post](/blog/2025/09/12/kubernetes-v1-34-cri-cgroup-driver-lookup-now-ga/#announcement-kubernetes-is-deprecating-containerd-v1-y-support) or in [KEP-4033: Discover cgroup driver from CRI](https://kep.k8s.io/4033)

Reviewer 4: ```suggestion
You can find more in the [official blog post](/blog/2025/09/12/kubernetes-v1-34-cri-cgroup-driver-lookup-now-ga/#announcement-kubernetes-is-deprecating-containerd-v1-y-support) or in [KEP-4033: Discover cgroup driver from CRI](https://kep.k8s.io/4033).
```

Historically, administrators relied on manual "read/write loops"—often piping `kubectl get` into `kubectl replace—to` update schemas or re-encrypt data at rest. This method was inefficient and prone to conflicts, especially for large resources like Secrets. With this release, the built-in controller automatically handles update conflicts and consistency tokens, providing a safe, streamlined, and reliable way to ensure stored data remains current with minimal operational overhead.

This work was done as part of [KEP #4192](https://kep.k8s.io/4192) led by SIG API Machinery.

### Mutable Volume attach limits

A CSI (Container Storage Interface) driver is a Kubernetes plugin that provides a consistent way for storage systems to be exposed to containerized workloads. The `CSINode` object records details about all CSI drivers installed on a node. However, a mismatch can arise between the reported and actual attachment capacity on nodes. When volume slots are consumed after a CSI driver starts up, the `kube-scheduler` may assign stateful pods to nodes without sufficient capacity, ultimately getting stuck in a `ContainerCreating` state.

> Kubernetes v1.35 makes `CSINode.spec.drivers[*].allocatable.count` mutable so that a node’s available volume attachment capacity can be updated dynamically. It also allows CSI drivers to control how frequently the `allocatable.count` value is updated on all nodes by introducing a configurable refresh interval, defined through the `CSIDriver` object. Additionally, it automatically updates `CSINode.spec.drivers[*].allocatable.count` on detecting a failure in volume attachment due to insufficient capacity. Although this feature graduated to Beta in v1.34 with the feature flag `MutableCSINodeAllocatableCount` disabled by default, it remains in Beta for v1.35 to allow time for feedback, but the feature flag is enabled by default.

Author: done

This work was done as part of [KEP #4639](https://kep.k8s.io/4639) led by SIG Node.

### Enforced `kubelet` credential verification for cached images

The `imagePullPolicy: IfNotPresent` setting currently allows a Pod to use a container image that is already cached on a node, even if the Pod itself does not possess the credentials to pull that image. A drawback of this behavior is that it creates a security vulnerability in multi-tenant clusters: if a Pod with valid credentials pulls a sensitive private image to a node, a subsequent unauthorized Pod on the same node can access that image simply by relying on the local cache.

This KEP introduces a mechanism where the `kubelet` enforces credential verification for cached images. Before allowing a Pod to use a locally cached image, the `kubelet` checks if the Pod has the valid credentials to pull it. This ensures that only authorized workloads can use private images, regardless of whether they are already present on the node, significantly hardening the security posture for shared clusters.

> In Kubernetes v1.35, this feature has graduated to Beta and is enabled by default. Users can still disable it by setting the `KubeletEnsureSecretPulledImages` feature gate to false. Additionally, the `imagePullCredentialsVerificationPolicy` flag allows operators to configure the desired security level, ranging from a mode that prioritizes backward compatibility to a strict enforcement mode that offers maximum security.

Author: done

### Gang scheduling support in Kubernetes

Scheduling interdependent workloads, such as AI/ML training jobs or HPC simulations, has traditionally been challenging because the default Kubernetes scheduler places Pods individually. This often leads to partial scheduling where some Pods start while others wait indefinitely for resources, resulting in deadlocks and wasted cluster capacity.

Kubernetes v1.35 introduces native Gang Scheduling support via the new Workload API and PodGroup concept. This feature implements an "all-or-nothing" scheduling strategy, ensuring that a defined group of Pods is scheduled only if the cluster has sufficient resources to accommodate the entire group simultaneously.

The primary benefit is improved reliability and efficiency for batch and parallel workloads. By preventing partial deployments, it eliminates resource deadlocks and ensures that expensive cluster capacity is utilized only when a complete job can run, significantly optimizing the orchestration of large-scale data processing tasks.

This work was done as part of [KEP #4671](https://kep.k8s.io/4671) led by SIG Scheduling.

### Continued innovation in Dynamic Resource Allocation (DRA)

> The [core functionality](https://kep.k8s.io/4381) was graduated to stable in v1.34, with the ability to turn it off. In v1.35 it is always enabled. Several alpha features have also been significantly improved and are ready for testing. We encourage users to provide feedback on these capabilities to help clear the path for their target promotion to beta in upcoming releases.

Author: I think it's clear, we should just move it to "Other notable changes section"

* [Kubelet limit of Parallel Image Pulls](https://kep.k8s.io/3673)
* [Comparable Resource Version](https://kep.k8s.io/5504)
* [Add a TopologyManager policy option for MaxAllowableNUMANodes](https://kep.k8s.io/4622)
* [Include kubectl command metadata in http request headers](https://kep.k8s.io/859)
* [PreferSameNode Traffic Distribution (formerly PreferLocal traffic policy / Node-level topology)](https://kep.k8s.io/3015)
* [Job API managed-by mechanism](https://kep.k8s.io/4368)
* [Transition from SPDY to WebSockets](https://kep.k8s.io/4006)

### Deprecations, removals and community updates

As Kubernetes develops and matures, features may be deprecated, removed, or replaced with better
ones to improve the project's overall health. See the Kubernetes
[deprecation and removal policy](/docs/reference/using-api/deprecation-policy/) for more details on this process. Kubernetes v1.35 includes a couple of deprecations.

#### Ingress NGINX retirement

For years, the Ingress NGINX controller has been a popular choice for routing traffic into Kubernetes clusters. It was flexible, widely adopted, and served as the standard entry point for countless applications. 

> However, maintaining the project has become unsustainable. With a severe shortage of maintainers and mounting technical debt the community recently made the difficult decision to retire it. This isn't strictly part of the v1.35 release, but it's such an important change that we wanted to call it out anyway.

Reviewer 4: ```suggestion
However, maintaining the project has become unsustainable. With a severe shortage of maintainers and mounting technical debt the community recently made the difficult decision to retire it. This isn't strictly part of the v1.35 release, but it's such an important change that we wanted to highlight it here.
```

Consequently, the Kubernetes project announced that Ingress NGINX will receive only best-effort maintenance until **March 2026**. After this date, it will be archived with no further updates. The recommended path forward is to migrate to the [Gateway API](https://gateway-api.sigs.k8s.io/), which offers a more modern, secure, and extensible standard for traffic management.

You can find more in the [official blog post](/blog/2025/11/11/ingress-nginx-retirement/).

> #### cgroup v1 support

Reviewer 4: ```suggestion
#### Removal of cgroup v1 support
```

> When it comes to managing resources on Linux nodes, Kubernetes has historically relied on cgroups (control groups). While the original cgroup v1 got the job done, it was often inconsistent and limited. That is why Kubernetes introduced support for cgroup v2 back in v1.25, offering a much cleaner, unified hierarchy and better resource isolation.

Reviewer 4: ```suggestion
When it comes to managing resources on Linux nodes, Kubernetes has historically relied on cgroups (control groups). While the original cgroup v1 was functional, it was often inconsistent and limited. That is why Kubernetes introduced support for cgroup v2 back in v1.25, offering a much cleaner, unified hierarchy and better resource isolation.
```

Author: done

Years ago, Kubernetes adopted the [`ipvs`](/docs/reference/networking/virtual-ips/#proxy-mode-ipvs) mode in `kube-proxy` to provide faster load balancing than the standard [`iptables`](/docs/reference/networking/virtual-ips/#proxy-mode-iptables). While it offered a performance boost, keeping it in sync with evolving networking requirements created too much technical debt and complexity.

Because of this maintenance burden, Kubernetes v1.35 deprecates `ipvs` mode. Although the mode remains available in this release, `kube-proxy` will now emit a warning on startup when configured to use it. The goal is to streamline the codebase and focus on modern standards. For Linux nodes, you should begin transitioning to [`nftables`](/docs/reference/networking/virtual-ips/#proxy-mode-nftables), which is now the recommended replacement.

You can find more in [KEP-5495: Deprecate ipvs mode in kube-proxy](https://kep.k8s.io/5495)

#### Final call for containerd v1.X

While Kubernetes v1.35 still supports containerd 1.7 and other LTS releases, this is effectively the end of the road. The SIG Node community has designated v1.35 as the last release to support the containerd v1.X series.

> This serves as a final warning: before upgrading to the next Kubernetes version, you must switch to containerd 2.0 or later. To help identify which nodes need attention, you can monitor the `kubelet_cri_losing_support` metric within your cluster.

Reviewer 4: ```suggestion
While Kubernetes v1.35 still supports containerd 1.7 and other LTS releases, this is the final version with such support. The SIG Node community has designated v1.35 as the last release to support the containerd v1.X series.

This serves as an important reminder: before upgrading to the next Kubernetes version, you must switch to containerd 2.0 or later. To help identify which nodes need attention, you can monitor the `kubelet_cri_losing_support` metric within your cluster.
```
